File: app/src/main/kotlin/com/gtaclone/engine/InputManager.kt
package com.gtaclone.engine

import android.view.MotionEvent
import com.gtaclone.ui.VirtualJoystick
import com.gtaclone.world.World

class InputManager {

    var joystickPointerId = -1

    fun handleTouch(event: MotionEvent, joystick: VirtualJoystick, world: World, screenW: Float, screenH: Float) {
        val actionIndex = event.actionIndex
        val pointerId = event.getPointerId(actionIndex)
        val x = event.getX(actionIndex)
        val y = event.getY(actionIndex)

        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN, MotionEvent.ACTION_POINTER_DOWN -> {
                if (joystick.contains(x, y) && joystickPointerId == -1) {
                    joystickPointerId = pointerId
                    joystick.onTouchDown(x, y)
                } else {
                    handleButton(x, y, screenW, screenH, world)
                }
            }
            MotionEvent.ACTION_MOVE -> {
                for (i in 0 until event.pointerCount) {
                    if (event.getPointerId(i) == joystickPointerId) {
                        joystick.onTouchMove(event.getX(i), event.getY(i))
                    }
                }
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_POINTER_UP -> {
                if (pointerId == joystickPointerId) {
                    joystickPointerId = -1
                    joystick.onTouchUp()
                }
            }
        }
    }

    private fun handleButton(x: Float, y: Float, screenW: Float, screenH: Float, world: World) {
        val btnSize = screenH * 0.11f
        val margin = 20f
        val rightX = screenW - margin - btnSize
        val bottomY = screenH - margin - btnSize

        // Car button (enter/exit vehicle)
        if (x >= rightX - btnSize * 1.3f && x <= rightX - btnSize * 0.3f &&
            y >= bottomY - btnSize * 1.3f && y <= bottomY - btnSize * 0.3f) {
            world.player.tryEnterExitVehicle(world)
            return
        }
        // Attack button
        if (x >= rightX && x <= rightX + btnSize && y >= bottomY && y <= bottomY + btnSize) {
            world.player.attack(world)
            return
        }
        // Sprint toggle
        if (x >= rightX && x <= rightX + btnSize &&
            y >= bottomY - btnSize * 1.3f && y <= bottomY - btnSize * 0.3f) {
            world.player.sprinting = !world.player.sprinting
        }
    }
}

File: GTAClone-iOS/GTAClone/engine/InputManager.swift
import UIKit

class InputManager {

    private weak var joystickTouch: UITouch?
    private weak var orbitTouch: UITouch?
    private var lastOrbitX: CGFloat = 0

    func began(_ touches: Set<UITouch>, world: World3D, joystick: VirtualJoystick, bounds: CGRect) {
        for t in touches {
            let p = t.location(in: nil)
            if joystick.contains(p) && joystickTouch == nil {
                joystickTouch = t; joystick.onTouchDown(p)
            } else if p.x > bounds.width * 0.55 {
                handleButton(p, bounds: bounds, world: world)
                if orbitTouch == nil { orbitTouch = t; lastOrbitX = p.x }
            }
        }
    }

    func moved(_ touches: Set<UITouch>, world: World3D, joystick: VirtualJoystick) {
        for t in touches {
            let p = t.location(in: nil)
            if t === joystickTouch { joystick.onTouchMove(p) }
            if t === orbitTouch {
                world.camera.orbitOffset += Float(p.x - lastOrbitX) * 0.3
                lastOrbitX = p.x
            }
        }
    }

    func ended(_ touches: Set<UITouch>, joystick: VirtualJoystick) {
        for t in touches {
            if t === joystickTouch { joystickTouch = nil; joystick.onTouchUp() }
            if t === orbitTouch    { orbitTouch = nil }
        }
    }

    private func handleButton(_ p: CGPoint, bounds: CGRect, world: World3D) {
        let bs = bounds.height * 0.11
        let mg: CGFloat = 20
        let rx = bounds.width  - mg - bs
        let by = bounds.height - mg - bs

        if p.x >= rx - bs*1.3 && p.x <= rx - bs*0.3 && p.y >= by - bs*1.3 && p.y <= by - bs*0.3 {
            world.player.tryEnterExitVehicle(world: world)
        }
        if p.x >= rx && p.x <= rx+bs && p.y >= by && p.y <= by+bs {
            world.player.attack(world: world)
        }
        if p.x >= rx && p.x <= rx+bs && p.y >= by-bs*1.3 && p.y <= by-bs*0.3 {
            world.player.sprinting.toggle()
        }
    }
}
